mesonbuild/dependencies/detect.py

```diff
@@ -108,13 +108,20 @@ def find_external_dependency(name: str, env: 'Environment', kwargs: T.Dict[str,
 
     for c in candidates:
         # try this dependency method
+        d: T.Optional[ExternalDependency] = None
         try:
             d = c()
             d._check_version()
             pkgdep.append(d)
         except DependencyException as e:
             assert isinstance(c, functools.partial), 'for mypy'
-            bettermsg = f'Dependency lookup for {name} with method {c.func.log_tried()!r} failed: {e}'
+            if d is not None:
+                method = f"with method '{d.log_tried()}'"
+            elif isinstance(c.func, ExternalDependency):
+                method = f"with method '{c.func.log_tried()}'"
+            else:
+                method = 'unable to determine method'
+            bettermsg = f'Dependency lookup for {name} {method!r} failed: {e}'
```

Comment:
It could be a `functools.partial`, it could be a `lambda`, it could be a generic function, the only requirement we enforce is that it has to be a function with the signature `def _() -> ExternalDependency`

Reply:
I've updated the message to be a little clearer. This *should* be a really hard error to hit, though clearly we *are* hitting it, which means we're having an exception in the `ExternalDependency` initializer. sigh

Reply:
I think it's really easy to have an exception in the initializer, because we run the entire class in the initializer and do all detection there. Including `raise DependencyException('cannot find Foobar because the stars do not align')`. This is an exception class that exists explicitly to be caught and safely handled as "abort the dependency lookup, log a reason, and report it as not found".

Of course we can sed all `def __init__` to `def run` and make lookups occur after initialization. Might take a couple more tweaks, I suppose.

Reply:
I don't know that it really helps. We probably should move the Exception throwing out of the ExternalDependency alltogether and have the helper function here check if `required` and `not_found`, but that's a big complicated set of changes.

Reply:
That would defeat the purpose of using exceptions to a) break out of the lookup, b) set a custom not-found message.

Using exceptions is idiomatic python for handling exceptional cases, aborting in `__init__` is the weird thing.

Reply:
Yeah, it's throwing in the initializer that's strange and makes things hard to handle, though really using exceptions for control flow isn't idiomatic in any language with exceptions, python included. But for the moment this at least gets rid of the traceback, and I'm not particularly interested in taking on the task of trying to clean up dependencies while I'm still neck deep in build.py...

Reply:
Can I reiterate that the short-term fix is still probably to fix the cmake openssl dependency, and not simply say "unknown"?

Reply:
The entire point of using a static method is that if the `ExternalDependency` is invalid, you can still get certain data about it. because of the exception that obviously doesn't work. So, the problem *is* that we hit an Exception in the initializer

Reply:
No, absolutely not at all. The exception has nothing to do with this. I already explained this several times.

Why would a staticmethod care if `__init__` throws an exception when per definition it operates on the class name before `__init__` is run?